### PR TITLE
USSD Complete registration

### PR DESCRIPTION
**Pull request by gsvr**

### CODE REVIEW

**Comment by gsvr** [comment]
@imsickofmaps @NicolaasDieBaas Please review.

File: src/ussd_registration.js [hunk 1]
```diff
@@ -134,10 +135,12 @@ go.app = function() {
             return new FreeText(name, {
                 question: $(questions[name]),
                 check: function(content) {
+                    var personnel_code = content;
                     return go.utils
-                        .validate_personnel_code(self.im, content)
-                        .then(function(valid_personnel_code) {
-                            if (valid_personnel_code) {
+                        .find_nurse_with_personnel_code(self.im, personnel_code)
```
Comment (imsickofmaps) [inline src/ussd_registration.js]: please replace `nurse` refs with `healthworker`

File: src/utils.js [hunk 2]
```diff
@@ -711,5 +765,47 @@ go.utils = {
         return choices;
     },
 
+    save_identities: function(im, receiver, receiver_msisdn, father_msisdn, mother_msisdn, operator_id) {
```
Comment (imsickofmaps) [inline src/utils.js]: These params need docs, I don't know if they're meant to be id store UUIDs or strings of MSISDNs or integers etc.

File: src/utils.js [hunk 3]
```diff
@@ -115,7 +115,7 @@ go.utils = {
     // Determine whether contact is registered
     is_registered: function(contact_id, im) {
         return go.utils
-            .get_contact_by_id(contact_id, im)
+            .get_identity_by_id(contact_id, im)
```
Comment (imsickofmaps) [inline src/utils.js]: PIN Number? :smile: 

Just `get_identity` is fine

File: src/utils.js [hunk 4]
```diff
@@ -203,25 +203,29 @@ go.utils = {
         }
     },
 
-// CONTACT HANDLING
+// IDENTITY HANDLING
 
-    get_contact_by_msisdn: function(msisdn, im) {
+    // Searches the Identity Store for all identities with the given msisdn.
+    // Returns the first identity object found
+    get_identity_by_msisdn: function(msisdn, im) {
```
Comment (imsickofmaps) [inline src/utils.js]: Please make `get_identity_by_address` and pass in a key: value pair

File: src/utils.js [hunk 5]
```diff
@@ -229,29 +233,42 @@ go.utils = {
             });
     },
 
-    // Create a new contact with the minimum required details
-    create_contact: function(msisdn, im) {
-        var payload = {
-            "details": {
+    // Create a new identity
+    create_identity: function(im, msisdn, communicate_through_id, operator_id) {
```
Comment (imsickofmaps) [inline src/utils.js]: same as above, address type and value

File: go-sms_inbound.js [hunk 6]
```diff
@@ -236,29 +240,42 @@ go.utils = {
...
-    create_contact: function(msisdn, im) {
-        var payload = {
-            "details": {
+    // Create a new identity
+    create_identity: function(im, msisdn, communicate_through_id, operator_id) {
+        var payload = {};
+
+        // compile base payload
+        if (msisdn) {
+            payload.details = {
```
Comment (imsickofmaps) [inline go-sms_inbound.js:258]: Just put the content of get_addresses in line here, it's not adding DRYness, just unnecessary jumping

File: go-sms_inbound.js [hunk 7]
```diff
@@ -236,29 +240,42 @@ go.utils = {
...
-                // Return the contact's id
-                return contact_created.id;
+                // Return the contact
+                return contact_created;
             });
     },
 
     // Gets a contact if it exists, otherwise creates a new one
-    get_or_create_contact: function(msisdn, im) {
+    get_or_create_identity: function(msisdn, im, operator_id) {
```
Comment (imsickofmaps) [inline go-sms_inbound.js]: lets make country code a config value and then passed into normalize function, else the value as a util drops

File: src/utils.js [hunk 8]
```diff
@@ -372,7 +389,9 @@ go.utils = {
 // OTHER
 
     get_addresses: function(msisdn) {
-        return "msisdn:" + msisdn;
+        var addresses = {"msisdn": {}};
```
Comment (imsickofmaps) [inline src/utils.js]: drop function

File: src/utils.js [hunk 9]
```diff
@@ -525,31 +544,31 @@ go.utils = {
...
-                return Q
-                    .all([
-                        // Update mama's contact
-                        go.utils.update_contact(im, mama_contact),
-                        // Create a subscription for mama
-                        go.utils.subscribe_contact(im, subscription)
-                    ]);
-            });
-    },
+    // save_contact_info_and_subscribe: function(im) {
```
Comment (imsickofmaps) [inline src/utils.js]: needs deleting?

**Comment by imsickofmaps** [comment]
@gsvr first pass done

**Comment by gsvr** [comment]
@imsickofmaps Second pass please.

**Comment by imsickofmaps** [comment]
:+1:

**Pull request merged**